plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.parquet;
+
+import io.trino.parquet.Field;
+import io.trino.parquet.reader.ParquetReader;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.LongArrayBlockBuilder;
+import io.trino.spi.type.Type;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A {@code ParquetPageSource} that inserts an column containing row indices
+ * into each page.
+ */
+public class RowIndexPageSource
+        extends ParquetPageSource
+{
+    private final ParquetReader reader;
+    private final int indexColumnIndex;
+
+    /**
+     * The first three parameters are inherited from and shared by the base
+     * {@link ParquetPageSource#ParquetPageSource(ParquetReader, List, List) ParquetPageSource}.
+     *
+     * @param indexColumnIndex The index at which the row-index column should be included.
+     */
+    public RowIndexPageSource(ParquetReader reader, List<Type> types, List<Optional<Field>> fields, int indexColumnIndex)
+    {
+        super(reader, types, fields);
+        this.reader = reader;
+        this.indexColumnIndex = indexColumnIndex;
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        Page page = super.getNextPage();
+        if (page == null) {
+            return null;
+        }
+
+        BlockBuilder builder = new LongArrayBlockBuilder(null, reader.lastBatchSize());
+        long end = reader.lastBatchStartRow() + reader.lastBatchSize();
+        for (long row = reader.lastBatchStartRow(); row < end; row++) {
+            builder.writeLong(row);
+        }
```

Comment:
This would look simpler:

```suggestion
        long[] rowIndexValues = new long[page.getPositionCount()];
        for (int position = 0; position < page.getPositionCount(); position++) {
            rowIndexValues[position] = reader.lastBatchStartRow() + position;
        }
        Block rowIndexBlock = new LongArrayBlock(page.getPositionCount(), Optional.empty(), rowIndexValues);
```

Reply:
(This code will disappear when I merge the page source classes.)